File: app/src/main/java/io/euphoria/xkcd/app/impl/ui/MessageForest.java
package io.euphoria.xkcd.app.impl.ui;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.NoSuchElementException;

/** Created by Xyzzy on 2020-03-19. */

public class MessageForest {

    public interface DisplayListener {

        void notifyItemRangeInserted(int start, int length);

        void notifyItemChanged(int index);

        void notifyItemMoved(int from, int to);

        void notifyItemRangeRemoved(int start, int length);

    }

    public static class DisplayListenerAdapter implements DisplayListener {

        private static DisplayListener NULL = new DisplayListenerAdapter();

        @Override
        public void notifyItemRangeInserted(int start, int length) {}

        @Override
        public void notifyItemChanged(int index) {}

        @Override
        public void notifyItemMoved(int from, int to) {}

        @Override
        public void notifyItemRangeRemoved(int start, int length) {}

    }

    /* All messages by ID. */
    private final Map<String, MessageTree> allMessages;
    /* The roots (i.e. messages with a parent of null) ordered by ID. */
    private final List<MessageTree> roots;
    /* Messages whose parent has not been added yet (mapping from parent ID). */
    private final Map<String, List<MessageTree>> orphans;
    /* The list of all visible messages in their proper order. */
    private final List<MessageTree> displayed;
    /* A listener for structural changes of the display list. */
    private DisplayListener listener;

    public MessageForest() {
        allMessages = new HashMap<>();
        roots = new ArrayList<>();
        orphans = new HashMap<>();
        displayed = new ArrayList<>();
        listener = DisplayListenerAdapter.NULL;
    }

    public DisplayListener getListener() {
        return listener;
    }

    public void setListener(DisplayListener listener) {
        if (listener == null) listener = DisplayListenerAdapter.NULL;
        this.listener = listener;
    }

    public int size() {
        return displayed.size();
    }

    public MessageTree get(int index) {
        return displayed.get(index);
    }

    public boolean has(String id) {
        return allMessages.containsKey(id);
    }

    public MessageTree get(String id) {
        return allMessages.get(id);
    }

    public MessageTree getParent(MessageTree mt) {
        return get(mt.getParent());
    }

    public MessageTree getThreadRoot(MessageTree mt) {
        while (true) {
            String parentID = mt.getParent();
            if (parentID == null) return mt;
            if (!has(parentID)) return null;
            mt = get(parentID);
        }
    }

    public MessageTree getReply(MessageTree mt, int index) {
        List<MessageTree> replies = mt.getReplies();
        if (index < 0) index += replies.size();
        if (index < 0 || index >= replies.size()) return null;
        return replies.get(index);
    }

    public MessageTree getSibling(MessageTree mt, int offset) {
        List<MessageTree> container;
        if (mt.getParent() == null) {
            container = roots;
        } else if (!has(mt.getParent())) {
            container = getOrphanList(mt.getParent());
        } else {
            container = getParent(mt).getReplies();
        }
        int index = Collections.binarySearch(container, mt);
        if (index < 0) throw new NoSuchElementException("Trying to get sibling of non-linked-in MessageTree " + mt);
        index += offset;
        if (index < 0 || index >= container.size()) return null;
        return container.get(index);
    }

    protected List<MessageTree> getOrphanList(String parentID) {
        List<MessageTree> ret = orphans.get(parentID);
        if (ret == null) {
            ret = new ArrayList<>();
            orphans.put(parentID, ret);
        }
        return ret;
    }

    protected int findRootDisplayIndex(MessageTree mt) {
        int index = 0;
        while (index < displayed.size()) {
            MessageTree cur = displayed.get(index);
            if (cur.compareTo(mt) >= 0) break;
            index += 1 + cur.countVisibleReplies();
        }
        return index;
    }

    protected int findDisplayIndex(MessageTree mt, boolean markUpdate) {
        if (mt.getParent() == null) return findRootDisplayIndex(mt);
        MessageTree parent = getParent(mt);
        if (parent == null || parent.isCollapsed()) return -1;
        int index = findDisplayIndex(parent, markUpdate);
        if (index == -1) return -1;
        if (markUpdate) listener.notifyItemChanged(index);
        index++;
        for (MessageTree sib : parent.getReplies()) {
            if (sib.compareTo(mt) >= 0) break;
            index += 1 + sib.countVisibleReplies();
        }
        return index;
    }

    public MessageTree add(MessageTree mt) {
        MessageTree existing = allMessages.get(mt.getID());
        if (existing == null) {
            processInsert(mt);
            return mt;
        } else if (!UIUtils.equalsOrNull(mt.getMessage(), existing.getMessage())) {
            processReplace(mt);
            return mt;
        } else {
            return existing;
        }
    }

    public MessageTree add(UIMessage msg) {
        MessageTree existing = allMessages.get(msg.getID());
        if (existing == null) {
            MessageTree mt = new MessageTree(msg);
            allMessages.put(mt.getID(), mt);
            processInsert(mt);
            return mt;
        } else if (!UIUtils.equalsOrNull(msg, existing.getMessage())) {
            existing.setMessage(msg);
            processReplace(existing);
            return existing;
        } else {
            return existing;
        }
    }

    public void setCollapsed(MessageTree mt, boolean collapsed) {
        if (collapsed == mt.isCollapsed()) return;
        processCollapse(mt, collapsed);
    }

    public void toggleCollapsed(MessageTree mt) {
        processCollapse(mt, !mt.isCollapsed());
    }

    public boolean tryEnsureVisible(MessageTree mt) {
        MessageTree parent = getParent(mt);
        if (mt.getParent() == null) {
            // Roots are always visible.
            return true;
        } else if (parent == null) {
            // Orphans cannot be made visible.
            return false;
        } else {
            // Otherwise, we go to the parent.
            setCollapsed(parent, false);
            return tryEnsureVisible(parent);
        }
    }

    public void move(MessageTree mt, MessageTree newParent, boolean ensureVisible) {
        if (ensureVisible && newParent != null) tryEnsureVisible(newParent);
        String newParentID = (newParent == null) ? null : newParent.getID();
        if (UIUtils.equalsOrNull(mt.getParent(), newParentID)) return;
        processMove(mt, newParent);
    }

    public MessageTree remove(MessageTree mt, boolean recursive) {
        MessageTree existing = allMessages.get(mt.getID());
        if (existing == null) return null;
        processRemove(existing, recursive);
        return existing;
    }

    protected void addDisplayRange(MessageTree mt, int index, boolean includeSelf) {
        List<MessageTree> toAdd = mt.traverseVisibleReplies(includeSelf);
        if (!includeSelf) index++;
        displayed.addAll(index, toAdd);
        listener.notifyItemRangeInserted(index, toAdd.size());
    }

    protected void removeDisplayRange(MessageTree mt, int index, boolean includeSelf) {
        int length = mt.countVisibleReplies();
        if (includeSelf) {
            length++;
        } else {
            index++;
        }
        displayed.subList(index, index + length).clear();
        listener.notifyItemRangeRemoved(index, length);
    }

    protected void notifyItemMovedLenient(int from, int to) {
        if (from == -1 && to == -1) {
            /* NOP */
        } else if (from == -1) {
            listener.notifyItemRangeInserted(to, 1);
        } else if (to == -1) {
            listener.notifyItemRangeRemoved(from, 1);
        } else {
            listener.notifyItemMoved(from, to);
        }
    }

    protected void processInsert(MessageTree mt) {
        allMessages.put(mt.getID(), mt);
        // Adopt orphans! :)
        List<MessageTree> children = orphans.remove(mt.getID());
        if (children != null) mt.addReplies(children);
        // The next steps depend on whether the message has a parent.
        int displayIndex;
        String parentID = mt.getParent();
        if (parentID == null) {
            // If there is no parent, this is a new root.
            UIUtils.insertSorted(roots, mt);
            displayIndex = findRootDisplayIndex(mt);
        } else if (!has(parentID)) {
            // If the parent does not exist, the message is an orphan.
            getOrphanList(parentID).add(mt);
            return;
        } else {
            // Otherwise, a parent exists.
            displayIndex = findDisplayIndex(mt, true);
            get(parentID).addReply(mt);
            // Updating the display list does, naturally, not happen for invisible messages.
            if (displayIndex == -1) return;
        }
        // Finally, splice the message (along with its replies!) into the display list.
        addDisplayRange(mt, displayIndex, true);
    }

    protected void processReplace(MessageTree mt) {
        // If an already-existing MessageTree's underlying message has been replaced, we only need to mark it as
        // changed (and do so for all its parents for good measure).
        int index = findDisplayIndex(mt, true);
        if (index != -1) listener.notifyItemChanged(index);
    }

    protected void processCollapse(MessageTree mt, boolean collapse) {
        // Locate the message, marking its parents for refreshing. If it is invisible, there is little to do.
        int displayIndex = findDisplayIndex(mt, true);
        if (displayIndex == -1) {
            mt.setCollapsed(collapse);
            return;
        }
        // Do not forget to mark the message itself for updating.
        listener.notifyItemChanged(displayIndex);
        // Now to the main branch.
        if (collapse) {
            removeDisplayRange(mt, displayIndex, false);
            mt.setCollapsed(true);
        } else {
            mt.setCollapsed(false);
            addDisplayRange(mt, displayIndex, false);
        }
    }

    protected void processMove(MessageTree mt, MessageTree newParent) {
        // The sequence of operations is somewhat tricky, in particular w.r.t. ensuring we pass the right indices
        // to the update listener.
        // First, unlink the message from the data structures.
        int oldIndex = findDisplayIndex(mt, true);
        if (oldIndex != -1) {
            displayed.remove(oldIndex);
        }
        MessageTree oldParent = getParent(mt);
        if (oldParent != null) {
            oldParent.removeReply(mt);
        } else if (mt.getParent() == null) {
            UIUtils.removeSorted(roots, mt);
        } else {
            List<MessageTree> siblings = orphans.get(mt.getParent());
            if (siblings != null) siblings.remove(mt);
        }
        // Now, flip the message's parent to the new one.
        mt.setParent((newParent == null) ? null : newParent.getID());
        // Next, re-link the message into the data structures.
        // Here, we cannot use findDisplayIndex()' auto-notification feature as we have not yet notified the listener
        // about the message's move.
        int newIndex = findDisplayIndex(mt, false);
        if (newIndex != -1) {
            displayed.add(newIndex, mt);
        }
        if (newParent != null) {
            newParent.addReply(mt);
        } else {
            UIUtils.insertSorted(roots, mt);
        }
        // Finally, issue listener notifications.
        notifyItemMovedLenient(oldIndex, newIndex);
        findDisplayIndex(mt, true);
    }

    protected void processRemove(MessageTree mt, boolean recursive) {
        allMessages.remove(mt.getID());
        if (recursive) {
            // If removing recursively, the children need not become orphans, but have to be removed from the ID index
            // instead.
            for (MessageTree r : mt.traverseVisibleReplies(false)) allMessages.remove(r.getID());
        } else {
            // Create orphans! :(
            getOrphanList(mt.getID()).addAll(mt.getReplies());
        }
        // Remove the message from the root list.
        if (mt.getParent() == null) UIUtils.removeSorted(roots, mt);
        // Locate the message in the display list, and mark its parents for updates.
        int displayIndex = findDisplayIndex(mt, true);
        if (displayIndex == -1) return;
        // Splice the message along with its replies out.
        removeDisplayRange(mt, displayIndex, true);
    }

}

File: app/src/main/java/io/euphoria/xkcd/app/impl/ui/data/MessageForest.kt
package io.euphoria.xkcd.app.impl.ui.data

import android.os.Parcel
import android.os.Parcelable
import io.euphoria.xkcd.app.impl.ui.UIUtils
import io.euphoria.xkcd.app.impl.ui.data.MessageTree
import java.util.*
import kotlin.collections.ArrayList
import kotlin.collections.HashMap

/** Created by Xyzzy on 2020-03-19.  */
class MessageForest() : Parcelable {
    /* All messages by ID. */
    private val allMessages: MutableMap<String, MessageTree> = HashMap()

    /* The roots (i.e. messages with a parent of null) ordered by ID. */
    private val roots: MutableList<MessageTree> = ArrayList()

    /* Messages whose parent has not been added yet (mapping from parent ID). */
    private val orphans: MutableMap<String, MutableList<MessageTree>> = HashMap()

    /* The list of all visible messages in their proper order. */
    private val displayed: MutableList<MessageTree> = ArrayList()

    /* A listener for structural changes of the display list. */
    private var _listener: DisplayListener = DisplayListenerAdapter.NULL
    val listener get() = _listener

    fun setListener(listener: DisplayListener?) {
        this._listener = listener ?: DisplayListenerAdapter.NULL
    }

    protected constructor(`in`: Parcel) : this() {
        roots.addAll(readGroupFromParcel(`in`))
        while (true) {
            val marker = `in`.readByte()
            if (marker.toInt() == 0) {
                break
            } else require(marker == MessageTree.Companion.PF_IS_A_THING) { "Invalid parcelled data" }
            readGroupFromParcel(`in`)
        }
        for (mt in roots) {
            displayed.addAll(mt.traverseVisibleReplies(true))
        }
    }

    override fun describeContents(): Int {
        return 0 // Nothing to see here.
    }

    override fun writeToParcel(out: Parcel, flags: Int) {
        writeGroupToParcel(out, null, roots)
        for ((key, value) in orphans) {
            out.writeByte(MessageTree.Companion.PF_IS_A_THING)
            writeGroupToParcel(out, key, value)
        }
        out.writeByte(0.toByte())
    }

    private fun addToAllMessagesRecursive(mt: MessageTree) {
        allMessages[mt.id] = mt
        for (r in mt.replies) {
            addToAllMessagesRecursive(r)
        }
    }

    private fun readGroupFromParcel(`in`: Parcel): List<MessageTree> {
        val parent = `in`.readString()
        val ret: MutableList<MessageTree> = ArrayList()
        while (true) {
            val flags = `in`.readByte()
            if (flags.toInt() == 0) break
            val mt = MessageTree(`in`, flags, parent)
            addToAllMessagesRecursive(mt)
            ret.add(mt)
        }
        if (parent != null) {
            orphans[parent] = ret
        }
        return ret
    }

    private fun writeGroupToParcel(
        out: Parcel,
        parent: String?,
        mts: List<MessageTree?>
    ) {
        out.writeString(parent)
        for (mt in mts) {
            mt!!.writeToParcel(out)
        }
        out.writeByte(0.toByte())
    }

    fun size(): Int {
        return displayed.size
    }

    operator fun get(index: Int): MessageTree? {
        return displayed[index]
    }

    fun indexOf(mt: MessageTree): Int {
        return findDisplayIndex(mt, visible = true, markUpdate = false)
    }

    fun has(id: String): Boolean {
        return allMessages.containsKey(id)
    }

    operator fun get(id: String): MessageTree? {
        return allMessages[id]
    }

    fun getParent(mt: MessageTree): MessageTree? {
        return mt.parent?.let { get(it) }
    }

    fun getThreadRoot(mt: MessageTree): MessageTree? {
        var mt = mt
        while (true) {
            val parentID = mt.parent ?: return mt
            mt = get(parentID) ?: return null
        }
    }

    fun getReply(mt: MessageTree, index: Int): MessageTree? {
        var index = index
        val replies = mt.replies
        if (index < 0) index += replies.size
        return if (index < 0 || index >= replies.size) null else replies[index]
    }

    fun getSibling(mt: MessageTree, offset: Int): MessageTree? {
        val container: List<MessageTree> = when {
            mt.parent == null -> roots
            !has(mt.parent!!) -> getOrphanList(mt.parent!!)
            else -> getParent(mt)!!.replies
        }

        var index = Collections.binarySearch(container, mt)
        if (index < 0) throw NoSuchElementException("Trying to get sibling of non-linked-in MessageTree $mt")
        index += offset
        return if (index < 0 || index >= container.size) null else container[index]
    }

    protected fun getOrphanList(parentID: String): MutableList<MessageTree> {
        var ret = orphans[parentID]
        if (ret == null) {
            ret = ArrayList()
            orphans[parentID] = ret
        }
        return ret
    }

    protected fun findRootDisplayIndex(mt: MessageTree, visible: Boolean): Int {
        var index = 0
        while (index < displayed.size) {
            val cur = displayed[index]
            if (cur >= mt) return if (visible && cur != mt) -1 else index
            index += 1 + cur.countVisibleReplies()
        }
        return if (visible) -1 else index
    }

    protected fun findDisplayIndex(
        mt: MessageTree,
        visible: Boolean,
        markUpdate: Boolean
    ): Int {
        if (mt.parent == null) return findRootDisplayIndex(mt, visible)
        val parent = getParent(mt)
        if (parent == null || parent.isCollapsed) return -1
        var index = findDisplayIndex(parent, false, markUpdate)
        if (index == -1) return -1
        if (markUpdate) _listener.notifyItemChanged(index)
        index++
        for (sib in parent.replies) {
            if (sib >= mt) return if (visible && sib != mt) -1 else index
            index += 1 + sib.countVisibleReplies()
        }
        return if (visible) -1 else index
    }

    fun clear() {
        val oldLength = displayed.size
        allMessages.clear()
        roots.clear()
        orphans.clear()
        displayed.clear()
        _listener.notifyItemRangeRemoved(0, oldLength)
    }

    fun add(mt: MessageTree): MessageTree? {
        val existing = allMessages[mt.id]
        return if (existing == null) {
            processInsert(mt)
            mt
        } else if (!UIUtils.equalsOrNull(mt.message, existing.message)) {
            processReplace(mt)
            mt
        } else {
            existing
        }
    }

    fun add(msg: UIMessage): MessageTree {
        val existing = allMessages[msg.id]
        return if (existing == null) {
            val mt = MessageTree(msg)
            processInsert(mt)
            mt
        } else if (!UIUtils.equalsOrNull(msg, existing.message)) {
            existing.setMessage(msg)
            processReplace(existing)
            existing
        } else {
            existing
        }
    }

    fun setCollapsed(mt: MessageTree, collapsed: Boolean) {
        if (collapsed == mt.isCollapsed) return
        processCollapse(mt, collapsed)
    }

    fun toggleCollapsed(mt: MessageTree) {
        processCollapse(mt, !mt.isCollapsed)
    }

    fun tryEnsureVisible(mt: MessageTree, expand: Boolean): Boolean {
        val ret: Boolean
        val parent = getParent(mt)
        ret = if (mt.parent == null) {
            // Roots are always visible.
            true
        } else if (parent == null) {
            // Orphans cannot be made visible.
            false
        } else {
            // Otherwise, we go to the parent.
            tryEnsureVisible(parent, true)
        }
        if (ret && expand) setCollapsed(mt, false)
        return ret
    }

    fun move(
        mt: MessageTree,
        newParent: MessageTree?,
        ensureVisible: Boolean
    ) {
        if (ensureVisible && newParent != null) tryEnsureVisible(newParent, true)
        val newParentID = newParent?.id
        if (UIUtils.equalsOrNull(mt.parent, newParentID)) return
        processMove(mt, newParent)
    }

    fun remove(mt: MessageTree, recursive: Boolean): MessageTree? {
        val existing = allMessages[mt.id] ?: return null
        processRemove(existing, recursive)
        return existing
    }

    protected fun addDisplayRange(
        mt: MessageTree,
        index: Int,
        includeSelf: Boolean
    ) {
        var index = index
        val toAdd = mt.traverseVisibleReplies(includeSelf)
        if (!includeSelf) index++
        displayed.addAll(index, toAdd)
        _listener.notifyItemRangeInserted(index, toAdd.size)
    }

    protected fun removeDisplayRange(
        mt: MessageTree,
        index: Int,
        includeSelf: Boolean
    ) {
        var index = index
        var length = mt.countVisibleReplies()
        if (includeSelf) {
            length++
        } else {
            index++
        }
        displayed.subList(index, index + length).clear()
        _listener.notifyItemRangeRemoved(index, length)
    }

    protected fun processInsert(mt: MessageTree) {
        allMessages[mt.id] = mt
        // Adopt orphans! :)
        val children: List<MessageTree>? = orphans.remove(mt.id)
        if (children != null) mt.addReplies(children)
        // The next steps depend on whether the message has a parent.
        val displayIndex: Int
        val parentID = mt.parent
        if (parentID == null) {
            // If there is no parent, this is a new root.
            UIUtils.insertSorted(roots, mt)
            mt.updateIndent(0)
            displayIndex = findRootDisplayIndex(mt, visible = false)
        } else if (!has(parentID)) {
            // If the parent does not exist, the message is an orphan.
            getOrphanList(parentID).add(mt)
            return
        } else {
            // Otherwise, a parent exists.
            displayIndex = findDisplayIndex(mt, visible = false, markUpdate = true)
            get(parentID)!!.addReply(mt)
            // Updating the display list does, naturally, not happen for invisible messages.
            if (displayIndex == -1) return
        }
        // Finally, splice the message (along with its replies!) into the display list.
        addDisplayRange(mt, displayIndex, includeSelf = true)
    }

    protected fun processReplace(mt: MessageTree) {
        // If an already-existing MessageTree's underlying message has been replaced, we only need to mark it as
        // changed (and do so for all its parents for good measure).
        val index = findDisplayIndex(mt, visible = true, markUpdate = true)
        if (index != -1) _listener.notifyItemChanged(index)
    }

    protected fun processCollapse(mt: MessageTree, collapse: Boolean) {
        // Locate the message, marking its parents for refreshing. If it is invisible, there is little to do.
        val displayIndex = findDisplayIndex(mt, visible = true, markUpdate = true)
        if (displayIndex == -1) {
            mt.isCollapsed = collapse
            return
        }
        // Do not forget to mark the message itself for updating.
        _listener.notifyItemChanged(displayIndex)
        // Now to the main branch.
        if (collapse) {
            removeDisplayRange(mt, displayIndex, includeSelf = false)
            mt.isCollapsed = true
        } else {
            mt.isCollapsed = false
            addDisplayRange(mt, displayIndex, includeSelf = false)
        }
    }

    protected fun processMove(mt: MessageTree, newParent: MessageTree?) {
        // Just in case someone "moves" a nonexistent message.
        if (!allMessages.containsKey(mt.id)) allMessages[mt.id] = mt
        // The sequence of operations is somewhat tricky, in particular w.r.t. ensuring we pass the right indices
        // to the update listener.
        // First, unlink the message from the data structures.
        val oldIndex = findDisplayIndex(mt, visible = true, markUpdate = true)
        if (oldIndex != -1) {
            displayed.removeAt(oldIndex)
        }
        val oldParent = getParent(mt)
        oldParent?.removeReply(mt)
            ?: if (mt.parent == null) {
                UIUtils.removeSorted(roots, mt)
            } else {
                val siblings =
                    orphans[mt.parent!!]
                siblings?.remove(mt)
            }
        // Now, flip the message's parent to the new one.
        mt.parent = newParent?.id
        // Next, re-link the message into the data structures.
        // Here, we cannot use findDisplayIndex()' auto-notification feature as we have not yet notified the listener
        // about the message's move.
        val newIndex = findDisplayIndex(mt, visible = false, markUpdate = false)
        if (newIndex != -1) {
            displayed.add(newIndex, mt)
        }
        if (newParent != null) {
            newParent.addReply(mt)
        } else {
            UIUtils.insertSorted(roots, mt)
            mt.updateIndent(0)
        }
        // Finally, issue listener notifications.
        DisplayListenerAdapter.Companion.notifyItemMovedLenient(_listener, oldIndex, newIndex)
        findDisplayIndex(mt, visible = true, markUpdate = true)
    }

    protected fun processRemove(mt: MessageTree, recursive: Boolean) {
        allMessages.remove(mt.id)
        if (recursive) {
            // If removing recursively, the children need not become orphans, but have to be removed from the ID index
            // instead.
            for (r in mt.traverseVisibleReplies(false)) allMessages.remove(r.id)
        } else {
            // Create orphans! :(
            getOrphanList(mt.id).addAll(mt.replies)
        }
        // Remove the message from the root list.
        if (mt.parent == null) UIUtils.removeSorted(roots, mt)
        // Locate the message in the display list, and mark its parents for updates.
        val displayIndex = findDisplayIndex(mt, true, true)
        if (displayIndex == -1) return
        // Splice the message along with its replies out.
        removeDisplayRange(mt, displayIndex, true)
    }

    companion object {
        val CREATOR: Parcelable.Creator<MessageForest?> =
            object : Parcelable.Creator<MessageForest?> {
                override fun createFromParcel(`in`: Parcel): MessageForest? {
                    return MessageForest(`in`)
                }

                override fun newArray(size: Int): Array<MessageForest?> {
                    return arrayOfNulls(size)
                }
            }
    }
}
